controllers/operatorscaler_controller.go

```diff
@@ -0,0 +1,333 @@
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=prometheuses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=alertmanagers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=odf.ibm.com,resources=flashsystemclusters,verbs=get;list;watch
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *OperatorScalerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+
+	r.log.Info("starting reconcile")
+
+	if err := r.isOdfDependenciesCsvReady(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileOperators(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileDynamicWatchers(req.Name); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("ending reconcile")
+	return ctrl.Result{}, nil
+}
+
+func (r *OperatorScalerReconciler) isOdfDependenciesCsvReady() error {
+
+	odfDepsCsv := &oprv1a1.ClusterServiceVersion{}
+
+	err := r.Client.Get(r.ctx, types.NamespacedName{Name: OdfDepsSubscriptionStartingCSV, Namespace: r.OperatorNamespace}, odfDepsCsv)
+	if err != nil {
+		r.log.Error(err, "failed getting odf-deps csv", "csvName", OdfDepsSubscriptionStartingCSV)
+		return err
+	}
+
+	if odfDepsCsv.Status.Phase != oprv1a1.CSVPhaseSucceeded {
+		err = fmt.Errorf("csv %s is not in succeeded state", OdfDepsSubscriptionStartingCSV)
+		r.log.Error(err, "waiting for csv to be in succeeded state")
+		return err
+	}
+
+	return nil
+}
+
+func (r *OperatorScalerReconciler) reconcileOperators() error {
+
+	var returnErr error
+
+	for _, metaData := range MetaDataList {
+
+		crList := &metav1.PartialObjectMetadataList{}
+		crList.TypeMeta.APIVersion = metaData.ApiVersion
+		crList.TypeMeta.Kind = metaData.Kind
+
+		err := r.Client.List(r.ctx, crList)
+		if err != nil {
+			if !meta.IsNoMatchError(err) {
+				msg := fmt.Sprintf("failed listing %s", metaData.Kind)
+				r.log.Error(err, msg)
+				multierr.AppendInto(&returnErr, err)
+			}
+		} else {
+			if len(crList.Items) >= 0 {
+
+				for _, pkgName := range metaData.PkgNames {
+					key := fmt.Sprintf(CsvLabelKey, pkgName, r.OperatorNamespace)
+
+					csvList := &oprv1a1.ClusterServiceVersionList{}
+					err = r.Client.List(
+						r.ctx, csvList,
+						client.InNamespace(r.OperatorNamespace),
+						client.MatchingLabels(map[string]string{key: ""}),
+					)
+					if err != nil {
+						r.log.Error(err, "failed listing csvs with label", "label", key)
+						multierr.AppendInto(&returnErr, err)
+					} else {
+						err = r.updateCsvsReplicas(csvList.Items...)
+						if err != nil {
+							r.log.Error(err, "failed updating csvs replica")
+							multierr.AppendInto(&returnErr, err)
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return returnErr
+}
+
+func (r *OperatorScalerReconciler) updateCsvsReplicas(csvList ...oprv1a1.ClusterServiceVersion) error {
```

Comment:
When it comes to code separation, it will be make more sense to have a func that handles a single CSV and update the deploment. That function then can be invoked multiple times (in a for loop) on the calling site
```suggestion
func (r *OperatorScalerReconciler) updateCsvDeplymentsReplicas(csv oprv1a1.ClusterServiceVersion) error {
```

Reply:
Done

Reply:
@iamniting You did not addressed the comment, you only renamed the function but the function still takes a list of CSVs

Reply:
@nb-ohad The way this func is written it can handle even a single entry, It will work for both of them.